pkg/maps/nat/stats/cell.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package stats
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cilium/cilium/pkg/maps/nat"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/index"
+	"github.com/cilium/hive/cell"
+	"github.com/spf13/pflag"
+)
+
+var Cell = cell.Module(
+	"nat-stats",
+	"Aggregates stats for NAT maps",
+	metrics.Metric(newMetrics),
+	cell.ProvidePrivate(newTables),
+	cell.Provide(
+		func(m Metrics) natMetrics {
+			return m
+		},
+		newStats,
+	),
+	cell.Config(Config{
+		NATMapStatInterval:       30 * time.Second,
+		NatMapStatKStoredEntries: 32,
+	}),
+	cell.Invoke(func(_ *Stats) {}),
+)
+
+type Metrics struct {
+	LocalPorts metric.Vec[metric.Gauge]
+}
+
+type natMetrics interface {
+	updateLocalPorts(family nat.IPFamily, count, maxPorts int)
+}
+
+func (m Metrics) updateLocalPorts(family nat.IPFamily, count, maxPorts int) {
+	m.LocalPorts.WithLabelValues(family.String()).Set(float64(count) / float64(maxPorts))
+}
+
+func newMetrics() Metrics {
+	return Metrics{
+		LocalPorts: metric.NewGaugeVecWithLabels(metric.GaugeOpts{
+			Namespace: metrics.Namespace,
+			Help:      "Max saturation of source ports on a egress-ip/external endpoint tuple",
+			Name:      "max_nat_endpoint_connections",
+		}, metric.Labels{
+			{Name: "ip", Values: metric.NewValues("ipv4", "ipv6")},
+		}),
+	}
+}
+
+type Config struct {
+	NATMapStatInterval       time.Duration `mapstructure:"nat-map-stats-interval"`
+	NatMapStatKStoredEntries int           `mapstructure:"nat-map-stats-entries"`
+}
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Duration("nat-map-stats-interval", def.NATMapStatInterval, "Interval upon which nat maps are iterated for stats")
+	flags.Int("nat-map-stats-entries", def.NatMapStatKStoredEntries, "Number k top stats entries to store locally in statedb")
+	flags.MarkHidden("nat-map-stats-entries")
+}
+
+var (
+	Index = statedb.Index[NatMapStats, string]{
+		Name: "byTuple",
+		FromObject: func(s NatMapStats) index.KeySet {
+			return index.StringSlice(
+				[]string{fmt.Sprintf("%s/%s %s %s:%d", s.Type, s.Proto, s.EgressIP, s.EndpointIP, s.RemotePort)})
```

Comment:
I'm not sure if this is a "hotpath" function, but `fmt.Sprintf` can be quite inefficient and expensive. This seems like a straightforward string concat, so if it's hotpath function, probably worth using `+` instead.

Reply:
good point! thanks